close.go

```diff
@@ -18,6 +18,12 @@ type graceful struct{}
 func (t graceful) shutdown(ctx context.Context, session *mgo.Session, closedChannel chan bool) {
 	session.Close()
 
+	defer func() {
+		if x := recover(); x != nil {
+			// do nothing ... just handle timing corner case and avoid "panic: send on closed channel"
```

Comment:
Hello Nathan and David,

As the code is in the middle of an "INTENTIONAL" shutdown() sequence I don't know if there is anything to be gained from logging that an some odd combination of timing has induced a panic().

The point of recovering from this 'corner case' panic() is purely to ensure that a graceful shutdown() is completed ... so that should the code be further evolved in the future any subsequent intentional shutdown actions get their opportunity to complete.

However, as I am not running your full suite of applications / microservices, I can not know the bigger picture and logging this particular may well be beneficail in hunting down other problems.

It would not do any harm, but might just waste peoples time in looking at a panic log(s) that are not really a problem - but just possibly an anomaly.

Kind Regards,

Rhys

Reply:
Hi Rhys,

after looking into the reason for a panic here (closing an already closed channel), I agree it doesn't add a lot of value and it would be better not to add another dependency to the library in log.go.

The user case is to continue the graceful shutdown in our applications and so if the channel is closed that is fine and we must continue (not block or exit early due to panic in library code).

Thanks

Nathan